fe/fe-core/src/main/java/com/starrocks/sql/spm/SPMPlanner.java

```diff
@@ -0,0 +1,240 @@
+            Preconditions.checkNotNull(node.getFn());
+            Preconditions.checkNotNull(other.getFn());
+            if (!StringUtils.equals(node.getFn().functionName(), other.getFn().functionName())) {
+                return false;
+            }
+            return check(node.getChildren(), ((Expr) node2).getChildren());
+        }
+        // ----------------- end -----------------
+
+        @Override
+        public Boolean visitInPredicate(InPredicate node, ParseNode context) {
+            if (node.getChildren().stream().noneMatch(SPMFunctions::isSPMFunctions)) {
+                return super.visitExpression(node, context);
+            }
+            InPredicate other = cast(context);
+            if (node.isNotIn() != other.isNotIn() || !check(node.getChild(0), other.getChild(0))) {
+                return false;
+            }
+            Preconditions.checkState(node.getChildren().size() == 2);
+            Preconditions.checkState(SPMFunctions.isSPMFunctions(node.getChild(1)));
+            Preconditions.checkState(!node.getChild(1).getChildren().isEmpty());
+            Preconditions.checkState(node.getChild(1).getChild(0) instanceof IntLiteral);
+            long spmId = ((IntLiteral) node.getChild(1).getChild(0)).getValue();
+            if (placeholderValues.containsKey(spmId)) {
+                // same placeholder check is same values
+                return placeholderValues.get(spmId).equals(other);
+            }
+            placeholderValues.put(spmId, other);
+            return true;
+        }
+    }
+
+    private class PlaceholderReplacer extends SPMUpdateExprVisitor<Void> {
+        @Override
+        public ParseNode visitFunctionCall(FunctionCallExpr node, Void context) {
+            if (SPMFunctions.isSPMFunctions(node)) {
+                Preconditions.checkState(!node.getChildren().isEmpty());
+                Preconditions.checkState(node.getChild(0) instanceof IntLiteral);
+                long id = ((IntLiteral) node.getChild(0)).getValue();
+                return placeholderValues.get(id);
+            }
+            return super.visitExpression(node, context);
+        }
+
+        @Override
+        public ParseNode visitInPredicate(InPredicate node, Void context) {
+            if (node.getChildren().stream().anyMatch(SPMFunctions::isSPMFunctions)) {
+                Preconditions.checkState(node.getChildren().size() == 2);
+                Preconditions.checkState(SPMFunctions.isSPMFunctions(node.getChild(1)));
+                Preconditions.checkState(!node.getChild(1).getChildren().isEmpty());
+                Preconditions.checkState(node.getChild(1).getChild(0) instanceof IntLiteral);
+                long spmId = ((IntLiteral) node.getChild(1).getChild(0)).getValue();
+                InPredicate value = placeholderValues.get(spmId).cast();
+                return new InPredicate(node.getChild(0), value.getListChildren(), node.isNotIn());
```

Comment:
why not return value directly?

Reply:
the `SlotExpr` transform depend on object, should use origin `SlotExpr`, don't replace it